src/Microsoft.Azure.WebJobs.Host/Queues/Listeners/QueueListener.cs

```diff
@@ -145,17 +146,36 @@ public void Dispose()
                 _stopWaitingTaskSource = new TaskCompletionSource<object>();
             }
 
-            IEnumerable<IStorageQueueMessage> batch;
+            IEnumerable<IStorageQueueMessage> batch  = null;
             try
             {
-                batch = await _queue.GetMessagesAsync(_queueProcessor.BatchSize,
-                    _visibilityTimeout,
-                    options: null,
-                    operationContext: null,
-                    cancellationToken: cancellationToken);
+                if (!_queueExists.HasValue || !_queueExists.Value)
+                {
+                    // Before querying the queue, determine if it exists. This
+                    // avoids generating unecessary exceptions (which pollute AppInsights logs)
+                    // Once we establish the queue exists, we won't do the existence
+                    // check anymore (steady state).
+                    // However the queue can always be deleted from underneath us, in which case
+                    // we need to recheck. That is handled below.
+                    _queueExists = await _queue.ExistsAsync(cancellationToken);
+                }
+
+                if (_queueExists.Value)
+                {
+                    batch = await _queue.GetMessagesAsync(_queueProcessor.BatchSize,
+                        _visibilityTimeout,
+                        options: null,
+                        operationContext: null,
+                        cancellationToken: cancellationToken);
+                }
             }
             catch (StorageException exception)
             {
+                // if we get ANY errors querying the queue reset our existence check
+                // doing this on all errors rather than trying to special case not
+                // found, because correctness is the most important thing here
+                _queueExists = null;
```

Comment:
A bit concerned with possible races with this...

Reply:
Can you elaborate? This entire method is called in a single threaded manner on a background timer.